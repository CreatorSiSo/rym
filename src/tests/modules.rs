use crate::{expr_parser, insta_assert_parser, item_parser, module_file_parser};
#[allow(unused_imports)]
use chumsky::Parser;
use indoc::indoc;

#[test]
fn empty() {
	insta_assert_parser! {
	item_parser(expr_parser());

	"mod name {}",
	"mod {}",

	"mod name",
	"mod name;",

	"mod",
	"mod;",
	}
}

#[test]
fn mixed() {
	insta_assert_parser! {
		item_parser(expr_parser());

		indoc! {r#"
		mod name {
			func testing() {}
			let this = testing();
		}"#
		},
	}
}

#[test]
fn file() {
	insta_assert_parser!(
		module_file_parser();

		indoc!(r#"
		/// useless function for adding two values together
		func add(x, y) {
			x + y; // here we add them
		}

		func multiply(x, y) {
			// here we are multiplying x by y
			x * y;
		}

		func main() {
			print("Hello World!");
			add(2, 3);
			multiply(3, 2);
		}"#)
	);
}

#[test]
fn recover_block() {
	insta_assert_parser! {
		item_parser(expr_parser());

		"mod out { let in_1 = 0; }",
		"mod out { let in_1 = 0; ]",

		"mod out { let in_1 = 0;   mod test   let in_1 = 0; }",
		"mod out { mod let in_1 = 0; }",

		"mod out { mod lvl_1 { mod lvl_2 { let in_3 = 0; }   mod lvl_2   let in_2 = 0; } }",
		"mod out { mod lvl_1 { mod lvl_2 { let in_3 = 0; ]   mod lvl_2   let in_2 = 0; } }", // only outer module is recovered
		"mod out { mod lvl_1 { mod lvl_2 { let in_3 = 0; )   mod lvl_2   let in_2 = 0; } }", // only outer module is recovered

		"mod name { mod name mod name {] }",
		"mod name { mod name; mod name {] }",
	}
}
