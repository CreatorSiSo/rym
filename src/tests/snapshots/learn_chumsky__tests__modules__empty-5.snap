---
source: src/tests/modules.rs
expression: "&snapshot"
---
--- Input ---
mod
---

Some(
    Spanned(
        Module {
            name: None,
            items: [],
        },
        0..3,
    ),
)

--- Errors ---
[
    ParseError {
        label: None,
        span: 1..3,
        reason: Unexpected,
        found: Eof,
        expected: [
            Semi,
            OpenBrace,
        ],
    },
]
---
